gotham/src/extractor/internal.rs

```diff
@@ -0,0 +1,1150 @@
+    }
+}
+
+impl de::Error for ExtractorError {
+    fn custom<T>(t: T) -> ExtractorError
+    where
+        T: Display,
+    {
+        ExtractorError::Custom(format!("{}", t))
+    }
+}
+
+/// Implements one `Deserializer` function (`$trait_fn`) to parse a single value using the
+/// `parse_single_value` function herein.
+macro_rules! single_value_type {
+    ($trait_fn:ident, $visitor_fn:ident) => {
+        fn $trait_fn<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+        where
+            V: Visitor<'de>
+        {
+            let v = parse_single_value(self.values)?;
+            visitor.$visitor_fn(v)
+        }
+    }
+}
+
+/// Implements one `Deserializer` function (`$trait_fn`) to return the error defined by the `$err`
+/// expression. For `Deserializer` functions with different signatures, the types that follow `self`
+/// can be provided as a trailing parameter list.
+macro_rules! reject_deserialize_type {
+    ($trait_fn:ident, $err:expr) => {
+        reject_deserialize_type!($trait_fn, $err, (_visitor: V));
+    };
+
+    {$trait_fn:ident, $err:expr, ($($arg_i:ident : $arg_t:ty),+)} => {
+        fn $trait_fn<V>(self, $($arg_i: $arg_t),+) -> Result<V::Value, Self::Error>
+        where V: Visitor<'de>
+        {
+            Err($err)
+        }
+    };
+}
+
+/// Specializes the `reject_deserialize_type` macro to return the `UnexpectedTargetType` variant,
+/// with the provided `$err` as the descriptive string.
+macro_rules! reject_target_type {
+    ($trait_fn:ident, $name:expr) => {
+        reject_target_type!($trait_fn, $name, (_visitor: V));
+    };
+
+    ($trait_fn:ident, $name:expr, ($($arg_i:ident : $arg_t:ty),+)) => {
+        reject_deserialize_type!(
+            $trait_fn,
+            ExtractorError::UnexpectedTargetType(
+                concat!("unsupported target type for path extractor: ", $name)
+            ),
+            ($($arg_i: $arg_t),+)
+        );
+    };
+}
+
+/// Specializes the `reject_deserialize_type` macro to return the `UnexpectedValueType` variant,
+/// with the provided `$err` as the descriptive string.
+macro_rules! reject_value_type {
+    ($trait_fn:ident, $name:expr) => {
+        reject_value_type!($trait_fn, $name, (_visitor: V));
+    };
+
+    ($trait_fn:ident, $name:expr, ($($arg_i:ident : $arg_t:ty),+)) => {
+        reject_deserialize_type!(
+            $trait_fn,
+            ExtractorError::UnexpectedValueType(
+                concat!("unsupported value type for path extractor: ", $name)
+            ),
+            ($($arg_i: $arg_t),+)
+        );
+    };
+}
+
+/// This trait represents the possible types that we can deserialize from when we're using
+/// extractors. The concrete values of this are all `IteratorAdaptor` types, and this trait is
+/// primarily giving us one place to deal with the type structure and expose the `next` function
+/// that we require.
+///
+/// See `from_segment_mapping` and TODO: `from_query_string` for how the values are constructed.
```

Comment:
minor nit: there are a few of these TODO scattered about, these appear to be resolved now, should they be cleaned up?

Reply:
Done.